File: go/best_time_to_buy_and_sell_stock.go
func maxProfit(prices []int) int {
	maxVal := 0
	minAmount := prices[0]

	for i := 1; i < len(prices); i++ {
		if minAmount > prices[i] {
			minAmount = prices[i]
			continue
		}
		if maxVal < prices[i]-minAmount {
			maxVal = prices[i] - minAmount
		}

	}

	return maxVal
}

File: ruby/best_time_to_buy_and_sell_stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
    max_profit = 0

    buy_point = prices[0]

    for i in 1...prices.length

        val = prices[i] - buy_point

        if buy_point>prices[i]
            buy_point = prices[i]
        end

        if max_profit < val

            max_profit = val

        end

    end


    return max_profit

end
